File: src/csharp/4_BehavioralPatterns/10_Strategy/Dynamic.cs
using System;
using System.Collections.Generic;
using System.Text;
using static System.Console;

namespace DesignPatterns
{
  public enum OutputFormat
  {
    Markdown,
    Html
  }

  public interface IListStrategy
  {
    void Start(StringBuilder sb);
    void End(StringBuilder sb);
    void AddListItem(StringBuilder sb, string item);
  }

  public class MarkdownListStrategy : IListStrategy
  {
    public void Start(StringBuilder sb)
    {
      // markdown doesn't require a list preamble
    }

    public void End(StringBuilder sb)
    {
      
    }

    public void AddListItem(StringBuilder sb, string item)
    {
      sb.AppendLine($" * {item}");
    }
  }

  public class HtmlListStrategy : IListStrategy
  {
    public void Start(StringBuilder sb)
    {
      sb.AppendLine("<ul>");
    }

    public void End(StringBuilder sb)
    {
      sb.AppendLine("</ul>");
    }

    public void AddListItem(StringBuilder sb, string item)
    {
      sb.AppendLine($"  <li>{item}</li>");
    }
  }

  public class TextProcessor
  {
    private StringBuilder sb = new StringBuilder();
    private IListStrategy listStrategy;

    public void SetOutputFormat(OutputFormat format)
    {
      switch (format) {
        case OutputFormat.Markdown:
          listStrategy = new MarkdownListStrategy();
          break;
        case OutputFormat.Html:
          listStrategy = new HtmlListStrategy();
          break;
        default:
          throw new ArgumentOutOfRangeException(nameof(format), format, null);
      }
    }

    public void AppendList(IEnumerable<string> items)
    {
      listStrategy.Start(sb);
      foreach (var item in items)
        listStrategy.AddListItem(sb, item);
      listStrategy.End(sb);
    }

    public StringBuilder Clear()
    {
      return sb.Clear();
    }

    public override string ToString()
    {
      return sb.ToString();
    }
  }

  class Demo
  { 
    static void Main(string[] args)
    {
      var tp = new TextProcessor();
      tp.SetOutputFormat(OutputFormat.Markdown);
      tp.AppendList(new []{"foo", "bar", "baz"});
      WriteLine(tp);

      tp.Clear();
      tp.SetOutputFormat(OutputFormat.Html);
      tp.AppendList(new[] { "foo", "bar", "baz" });
      WriteLine(tp);
    }
  }
}

File: src/swift/4_BehavioralPatterns/10_Strategy/Dynamic.swift
import Foundation
import XCTest

enum OutputFormat
{
  case markdown
  case html
}

protocol ListStrategy
{
  init()
  func start(_ buffer: inout String)
  func end(_ buffer: inout String)
  func addListItem(buffer: inout String, item: String)
}

class MarkdownListStrategy : ListStrategy
{
  required init() {}
  func start(_ buffer: inout String) {}
  func end(_ buffer: inout String) {}
  func addListItem(buffer: inout String, item: String)
  {
    buffer.append(" * \(item)\n")
  }
}

class HtmlListStrategy : ListStrategy
{
  required init() {}
  func start(_ buffer: inout String)
  {
    buffer.append("<ul>\n")
  }

  func end(_ buffer: inout String)
  {
    buffer.append("</ul>\n")
  }

  func addListItem(buffer: inout String, item: String)
  {
    buffer.append("  <li>\(item)</li>\n")
  }
}

class TextProcessor : CustomStringConvertible
{
  private var buffer = ""
  private var listStrategy: ListStrategy

  init(_ outputFormat: OutputFormat)
  {
    switch outputFormat
    {
    case .markdown:
      listStrategy = MarkdownListStrategy()
    case .html:
      listStrategy = HtmlListStrategy()
    }
  }

  func setOutputFormat(_ outputFormat: OutputFormat)
  {
    switch outputFormat
    {
    case .markdown:
      listStrategy = MarkdownListStrategy()
    case .html:
      listStrategy = HtmlListStrategy()
    }
  }

  func appendList(_ items: [String])
  {
    listStrategy.start(&buffer)
    for item in items
    {
      listStrategy.addListItem(buffer: &buffer, item: item)
    }
    listStrategy.end(&buffer)
  }

  func clear()
  {
    buffer = ""
  }

  var description: String
  {
    return buffer
  }
}

func main()
{
  let tp = TextProcessor(.markdown)
  tp.appendList(["foo", "bar", "baz"])
  print(tp.description)

  tp.clear()
  tp.setOutputFormat(.html)
  tp.appendList(["foo", "bar", "baz"])
  print(tp.description)
}

main()